inc/ustring.bi

```diff
@@ -0,0 +1,681 @@
+'' MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+'' ########################################################################################
+
+
+#IFDEF __FB_DOS__
+  TYPE DWSTR AS STRING
+  #define ustring STRING
+
+  #Print ======>
+  #Print ======> Warning, Compiled under DOS : DWSTR type is only an Alias for String!
+  #Print ======>
+
+#ELSE
+  #ifndef ustring
+    #ifdef afx                                        'if WinFBX is present
+      #ifndef cwstr                                   'if WinFBX is not used
+        #define ustring jk.DWSTR                      'use replacement
+
+      #else
+        #define ustring afx.CWSTR                     'use WinFBX�s dynamic Unicode string
+
+      #endif
+
+    #else
+      #define ustring jk.DWSTR                        'use replacement
+    #endif
+  #endif
+#endif
+
+
+#ifdef unicode   
+  #ifndef dstr                                        'dynamic unicode string
+    #define dstr ustring                              '(DSTR)
+  #endif  
+  #ifndef zstr                                        'zero terminated unicode string (* bytes)
+    #define zstr wstring
+  #endif  
+#ELSE
+  #ifndef dstr                                        'dynamic ansi string
+    #define dstr string
+  #endif  
+  #ifndef zstr                                        'zero terminated ansi string (* bytes)
+    #define zstr zstring
+  #endif  
+#ENDIF
+
+
+
+'***********************************************************************************************
+' replacement for Jos� Roca�s CWstr, if WinFBX is not present or is not used.
+' this is mostly a clone of his code without Windows dependency and adapted to
+' work wih Linux as well.
+'***********************************************************************************************
+
+
+#include once "crt\mem.bi"
+
+
+#IFNDEF __FB_DOS__
+
+
+NAMESPACE JK                                          'obviously i cannot use "AFX" here
+TYPE DWSTR_ AS DWSTR                                  'Forward reference
+
+
+'***********************************************************************************************
+' DWSTR CLASS
+'***********************************************************************************************
+
+#if sizeof(wstring) = 2
+TYPE DWSTR extends wstring
+  Private:
+    m_Capacity AS Ulong                               'the total size of the buffer
+    m_GrowSize AS LONG = 260 * SizeOf(WSTRING)        'how much to grow the buffer by when required
+
+  Public:
+    m_pBuffer AS UBYTE PTR                            'pointer to the buffer
+    m_BufferLen AS ulong                              'length in bytes of the current string in the buffer
+    u_size      as long = SizeOf(WSTRING)
+
+    DECLARE CONSTRUCTOR
+    DECLARE CONSTRUCTOR (BYVAL nChars AS ulong, BYVAL nCodePage AS ulong)
+    DECLARE CONSTRUCTOR (BYVAL pwszStr AS WSTRING PTR)
+    DECLARE CONSTRUCTOR (BYREF ansiStr AS STRING, BYVAL nCodePage AS ulong = 0)
```

Comment:
Probably should be `byref ansi as const string`.  Otherwise, it is not possible to initialize from a `const string`.  Applies to other declarations as well, as `const` not really used anywhere in this implementation.